lego/apps/events/tasks.py

```diff
@@ -137,64 +147,169 @@ def async_unregister(self, registration_id, logger_context=None):
+        log.error(
+            "Attempted to retrieve a non-existing payment intent",
+            registration=self.registration.id,
+        )
+        raise ValueError("The payment intent does not exist")
+
+    if client_secret is None:
+        try:
+            payment_intent = stripe.PaymentIntent.retrieve(
+                self.registration.payment_intent_id
+            )
+            client_secret = payment_intent["client_secret"]
+        except stripe.error.InvalidRequestError as e:
+            log.error(
+                "invalid_request", exception=e, registration_id=self.registration.id
+            )
+            self.registration.payment_status = e.json_body["error"]["type"]
+            self.registration.save()
+        except stripe.error.StripeError as e:
+            log.error("stripe_error", exception=e, registration_id=self.registration.id)
+            raise self.retry(exc=e, max_retries=3)
+        except stripe.error.APIConnectionError as e:
+            log.error(
+                "stripe_APIConnectionError",
+                exception=e,
+                registration_id=self.registration.id,
+            )
+            raise self.retry(exc=e, max_retries=3)
+        except Exception as e:
+            log.error(
+                "Exception on creating a payment intent",
+                exception=e,
+                registration=self.registration.id,
+            )
+            raise self.retry(exc=e)
+
+    notify_user_payment_initiated(
+        constants.SOCKET_INITIATE_PAYMENT_SUCCESS,
+        self.registration,
+        success_message="Betaling påbegynt",
+        client_secret=client_secret,
+    )
+
+
 @celery_app.task(base=Payment, bind=True)
-def async_payment(self, registration_id, token, logger_context=None):
+def async_initiate_payment(self, registration_id, logger_context=None):
+    """
+    Task that creates a new Stripe payment intent. Stripe returns a payment_intent object. The
+    client_secret is then needed to finsish the payment on the client side.
+    """
     self.setup_logger(logger_context)
 
     self.registration = Registration.objects.get(id=registration_id)
     event = self.registration.event
     try:
-        response = stripe.Charge.create(
+        payment_intent = stripe.PaymentIntent.create(
             amount=event.get_price(self.registration.user),
+            receipt_email=self.registration.user.email,
             currency="NOK",
-            source=token,
             description=event.slug,
+            idempotency_key=str(self.registration.id),
             metadata={
                 "EVENT_ID": event.id,
                 "USER_ID": self.registration.user.id,
                 "USER": self.registration.user.full_name,
                 "EMAIL": self.registration.user.email,
             },
         )
-        log.info("stripe_payment_success", registration_id=self.registration.id)
-        return response
-    except stripe.error.CardError as e:
```

Comment:
We don't handle any card errors explicitly anymore since all of the requests to Stripe that contain card data is handled on the client. The only payment error handling we do on the backend is monitoring the webhook event and setting the payment as failed, and the details are handled on the client.